docs/hooks.rst

```diff
@@ -4,44 +4,177 @@
-    }
-    // otherwise yields data row
-});
+    $m->breakHook(false);
 
-Also this approach can be used to prevent data row to be loaded. If you return false
+If you break beforeSave, then the save operation will not take place, although model will assume
+the operation was successful.
+
+You can also berak beforeLoad hook which can be used to skip rows::
+
+    $model->addHook('afterLoad', function ($m) {
+        if ($m['date'] < $m->date_from) {
+            $m->breakHook(false); // will not yield such data row
+        }
+        // otherwise yields data row
+    });
+
+This will also prevent data from being loaded. If you return false
 from afterLoad hook, then record which we just loaded will be instantly unloaded.
-This can be helpful in some cases.
+This can be helpful in some cases, although you should still use :php:meth:`Model::addCondition`
+where possible as it is much more efficient.
+
+Insert/Update Hooks
+-------------------
+
+Insert/Update are triggered from inside save() method but are based on current state of
+:php:meth:`Model::loaded`:
+
+ - beforeInsert($m, &$data) (creating new records only)
+ - afterInsert($m, $id)
+ - beforeUpdate($m, &$data) (updating existing records only. Not executed if model is not dirty)
+ - afterUpdate($m)
+
+The $data argument will contain array of actual data (field=>value) to be saved, which you can
+use to withdraw certain fields from actually being saved into the database (by unsetting it's value).
+
+afterInsert will receive either $id of new record or null if model couldn't procude ID field. Also,
+afterInsert is actually called before :php:meth:`Model::_reload_after_save` reloading is done.
+
+beforeSave, afterSave Hook
+--------------------------
+
+A good place to hook is beforeSave as it will be fired when adding new records or
+modifying existing ones:
+
+ - beforeSave($m) (saving existing or new records. Not executed if model is not dirty)
+ - afterSave($m) (same as above)
+
+You might consider "save" to be a higher level hook, as beforeSave is called pretty
+early on during saving the record and afterSave is caled at the very end of save.
+
+You may actually drop validation exception inside save, insert or update hooks::
+
+    $m->addHook('beforeSave', function($m) {
+        if ($m['name'] = 'Yagi') {
+            throw new \atk4\data\ValidationException(['name'=>"We don't serve like you"]);
+        }
+    });
+    
+Loading, Deleting
+-----------------
+
+Those are relatively simple hooks:
+
+ - beforeLoad($m, $id) ($m will be unloaded). Break for custom load or skip.
+ - afterLoad($m). ($m will contain data). Break to unload and skip.
+
+For the deletion it's pretty similar:
+
+ - beforeDelete($m, $id). Unload and Break to preserve record.
+ - afterDelete($m, $id).
+
+A good place to clean-up delete related records would be inside afterDelete, although
+if your database consistency requires those related records to be cleaned up first, use
+beforeDelete instead.
```

Comment:
Maybe small example how to use beforeDelete for soft-delete functionality would be good here?

Reply:
added link.